### ISSUE TITLE
Feature request: Correlation module

### ISSUE THREAD

**Issue by ClementPinard** [pytorch/pytorch#4073]
This module was used by FlowNetC, it is described [here](https://arxiv.org/pdf/1504.06852.pdf) (in section 3.)
Here is the caffe implementation of the authors
https://github.com/lmb-freiburg/flownet2/blob/master/src/caffe/layers/correlation_layer.cu
Here is one tensorflow implementation
https://github.com/sampepose/flownet2-tf/tree/master/src/ops/correlation

And also NVIDIA recently released a custom pytorch operator here
https://github.com/NVIDIA/flownet2-pytorch/tree/master/networks/correlation_package
here is another attempt in pytorch (but only high level)
https://github.com/onlytailei/pytorch-flownet2/blob/master/nets/flownetC.py#L45

Would it be useful to add such a feature in functional pytorch library ?

**Comment by kingsleyluoxin** [comment]
I have been wondering if anyone knows how to understand the codes of these self-defined operations.

**Comment by ClementPinard** [comment]
Basically, for each point in your output map, you compute the correlation of the two inputs, as if the patch of your `input2` was a kernel of convolution applied on patch of ̀input1`

other than padding there are 4 fundamental arguments to apply :
kernel size, displacement size, stride and dilation (for simplicity, we'll assume stride and dilation of 1 for now)

with a maximum displacement size, you would get this mapping :
``` Input1 BxCxHxW Input2 BxCxHxW -> Output BxHxWxHxW ```
which means you get to treat every possible patch of Input2 as normal convolution kernel, applied to the whole Input1
the minimum displacement you would get :
``` Output BxHxWx1x1 ```
The operation done is then just a bunch of inner products, for all `(n,i,j)` you compute the inner product of the two C-dimension vectors `<Input1[n,:,i,j], Input2[n,:,i,j]>` and then you apply a avgPool2d of your map with a stride of 1 and kernel_size of your correlation kernel size.

As all these operations are very similar to convolution, I have been studying on how we can make these operations simple GEMM calls, whether it is batched/strided or not (see [here](https://devblogs.nvidia.com/cublas-strided-batched-matrix-multiply/#more-7561) ), inspired by this presentation : http://www.prime-project.org/wp-content/uploads/sites/206/2018/02/Talk-10-David-Gregg-Parallel-Multi-Channel-Convolution-using-General-Matrix-Multiplication.pdf , but I cam across nothing conclusive, it seems we are bound to make batch GEMM call of very little matrices, and unable to benefit from CuBLAS capabilities.

I think the best options might be :
 - Do what the paper did with convolution, but instead of a GEMM primitive, we use a inner product primitive : construct every possible broadcast inner product of C-dimensional pixels (with spatial shift induced by displacement) followed by a avgpooling
 - Do what Fitsum Reda did in NVIDIA's repo : transpose both inputs to be `BxHxWxC` and then perform explicit correlation operation

I have not looked carefully in the backward operation yet, but it seems that the same kind of correlation operations will be done.

It seemed to me that Reda's technique was optimizable, especially  with the transposition and all the multiplications that were done multiple times across different correlation operation, but I don't have any proof that this is not the best way to deal with that very peculiar module

**Comment by ClementPinard** [comment]
Here is a pytorch way of seeing how can things be done, displacement diamter is D, kernel size is K, padding and stride are ignored for clarity

1st solution :
```python
d = (D-1)/2
for i in range(-d, d):
    for j in range(-d, d):
        output[:, i+ D*j, :, :] = torch.sum(input1 * input2[:, :, i:, j:], dim=1)  # [B, D^2, H, W]
final_output = F.avgpool2d(output, kernel_size = K, stride=1)  # padding ignored for clarity
```

2nd solution :
```python
rinput1 = input1.transpose(0,2,3,1)  # [B, H, W, C]
rinput2 = input2.transpose(0,2,3,1)  # [B, H, W, C]

for b in range(B):
    for i in range(H):
        for j in range(W):
            for k in range(-d, d):
                for l in range(-d, d):
                    # rinput1[b, i:i+K, j:j+K] is a patch of size [K, K, C]
                    output[b, k + D*l, i, j] = torch.sum(rinput1[b, i:i+K, j:j+K] * rinput2[b, i+k:i+k+K, j+l:j+l+K])
```

**Comment by ClementPinard** [comment]
As far as I understood, the key here is to operate on the most compact momery buffer possible. Here is a general summary of this idea : https://devblogs.nvidia.com/how-access-global-memory-efficiently-cuda-c-kernels/

So when you are doing the operation `torch.sum(rinput1[b, i:i+K, j:j+K] * rinput2[b, i+k:i+k+K, j+l:j+l+K]` in a cuda kernel, you need the strides of both patches to be the lowest possible.

When C is the last dimension, strides are W*C, C, 1 . When C is the second dimension, strides are H*W, W, 1
Now, depending on your number of channels, you might want to use one way of storing information of the other, but more generally, we have C < H, hence the transposition.

The weird thing however is that something similar to the `CUDA_KERNEL_LOOP` Macro is used (see here : https://devblogs.nvidia.com/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/) but only considering the number of channels. Why not do that directly with the index ? Is there a CUDA master able to explain this ?

we might have to ask Fitsum Reda and Alexey Dosovitskiy for this as I believe Dosovitskiy wrote the first correlation kernel in caffe and  Reda adapted it to pytorch, but without touching too much the inner mechanisms.

**Comment by kingsleyluoxin** [comment]
Thanks so much for your patient reply. I shall read and try to understand these codes with your explanation. :) 👍

**Comment by kingsleyluoxin** [comment]
Thank you so much for your implementation I shall learn it by heart.



发自我的小米手机
在 Clément Pinard <notifications@github.com>，2018年8月17日 下午6:25写道：

Here is a finished implementation, using the cpp-extension tools to make it installable.

https://github.com/ClementPinard/Pytorch-Correlation-extension

Note that this is still subject to change, as DeviceTensor is not an official interface. In the near future, it will most probably be replaced with what comes from #8366<https://github.com/pytorch/pytorch/issues/8366>

**Comment by pram991** [comment]
> 
> 
> Here is a pytorch way of seeing how can things be done, displacement diameter is D, kernel size is K, padding and stride are ignored for clarity
> Output is of size BxD^2xHxW but I think giving an output of size BxDxDxHxW would be semantically better if this module is to be used in other application than FlowNetC
> 
> 1st solution :
> 
> ```python
> d = (D-1)/2
> for i in range(-d, d):
>     for j in range(-d, d):
>         output[:, i+ D*j, :, :] = torch.sum(input1 * input2[:, :, i:, j:], dim=1)  # [B, D^2, H, W]
> final_output = F.avgpool2d(output, kernel_size = K, stride=1)  # padding ignored for clarity
> ```
> 
> 2nd solution :
> 
> ```python
> rinput1 = input1.transpose(0,2,3,1)  # [B, H, W, C]
> rinput2 = input2.transpose(0,2,3,1)  # [B, H, W, C]
> 
> for b in range(B):
>     for i in range(H):
>         for j in range(W):
>             for k in range(-d, d):
>                 for l in range(-d, d):
>                     # rinput1[b, i:i+K, j:j+K] is a patch of size [K, K, C]
>                     output[b, k + D*l, i, j] = torch.sum(rinput1[b, i:i+K, j:j+K] * rinput2[b, i+k:i+k+K, j+l:j+l+K])
> ```

How much slower would this solution be, compared to the cuda code in Pytorch-Correlation-extension?

@ClementPinard

**Comment by ClementPinard** [comment]
That would be interesting to test it with JIT tracing ! I haven't tested it, but there's a benchmark script in the Pytorch-Correlaton-extension repo that you can try to adapt to this python code.

My guess is that 1st will be slightly slower, but nothing too serious : I tried this solution in c++ with torch cpp and it was indeed a bit slower that the current implementation. However, 2nd solution has several parrallel computing tricks that makes it the fastest and it would be impossible to code it simply directly in pytorch, even with JIT tracing, so i would guess that it would be painfully slower (5 intricated for loops in python is hardly ever a good idea !)

So my pronostic, from fastest to slowest : 2nd solution (cpp/cuda) -> 1st solution (cpp/cuda) -> (by a short margin) 1st solution (python) -> 2nd solution (python)

**Comment by socome** [comment]
how can i use 1st solution in python ?? I don't

**Comment by bhack** [comment]
Any news on this?

**Comment by bhack** [comment]
After https://github.com/pytorch/pytorch/pull/127301 using @ClementPinard module in `torch.compile` is going to trigger:

```python 
/opt/conda/lib/python3.11/site-packages/torch/_dynamo/variables/functions.py:661: UserWarning: Graph break due to unsupported builtin spatial_correlation_sampler_backend.PyCapsule.forward. This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind). If it is a Python builtin, please file an issue on GitHub so the PyTorch team can add support for it and see the next case for a workaround. If it is a third-party C/C++ Python extension, please either wrap it into a PyTorch-understood custom operator (see https://pytorch.org/docs/main/notes/custom_operators.html for more details) or, if it is traceable, use torch.compiler.allow_in_graph```

**Comment by bhack** [comment]
@jansel @zou3519 as few days ago we merged https://github.com/pytorch/pytorch/pull/127301 what is the best approach to support this module?

**Comment by jansel** [comment]
Do you have a regular python version of this operator you could decompose the custom op to?  It might be worth comparing the perf of the decomposed version (and letting torch.compile fuse it with other things).

If not you can follow the instructions in the warning to wrap it in a custom op.

**Comment by bhack** [comment]
@jansel I don't think that we have a python version that we know it could decompose well. It would be ideal of course.

Also Meta research in some project is using de-facto the same third party implementation:
https://github.com/facebookresearch/NeRF-Det/blob/main/mmdet3d/mmcv/ops/csrc/pytorch/cuda/correlation_cuda.cu

**Comment by bhack** [comment]
@jansel There was a python only impl but with restricted args
https://github.com/limacv/CorrelationLayer

**Comment by jansel** [comment]
I'd recommend trying it out since it will be more fusible by compile.  If it is not faster, you can just wrap the other ones in a custom op.

For more on custom ops see:
https://pytorch.org/docs/main/notes/custom_operators.html

**Comment by bhack** [comment]
@jansel @ClementPinard  what do you think about this draft?

```python
class SpatialCorrelationSampler(nn.Module):
    def __init__(
        self,
        kernel_size=1,
        patch_size=1,
        stride=1,
        padding=0,
        dilation=1,
        dilation_patch=1,
    ):
        super(SpatialCorrelationSampler, self).__init__()
        self.kernel_size = _pair(kernel_size)
        self.patch_size = _pair(patch_size)
        self.stride = _pair(stride)
        self.padding = _pair(padding)
        self.dilation = _pair(dilation)
        self.dilation_patch = _pair(dilation_patch)
        self.max_displacement = patch_size // 2  # Correctly setting max_displacement
        self.padlayer = nn.ConstantPad2d(self.max_displacement, 0)

    @torch.compile(fullgraph=True)
    def forward(self, input1, input2):
        b, c, h, w = input1.shape

        # Pad the second input tensor
        input2_padded = self.padlayer(input2)

        # Unfold the input tensors to extract patches
        patches1 = F.unfold(
            input1,
            kernel_size=self.kernel_size,
            dilation=self.dilation,
            padding=self.padding,
            stride=self.stride,
        )

        # Calculate the output spatial dimensions
        output_height = (
            h + 2 * self.padding[0] - self.dilation[0] * (self.kernel_size[0] - 1) - 1
        ) // self.stride[0] + 1
        output_width = (
            w + 2 * self.padding[1] - self.dilation[1] * (self.kernel_size[1] - 1) - 1
        ) // self.stride[1] + 1

        patches1 = patches1.view(
            b,
            c * self.kernel_size[0] * self.kernel_size[1],
            output_height,
            output_width,
        )

        # Generate the offsets for the shifts
        offsets = torch.stack(
            torch.meshgrid(
                torch.arange(
                    -self.max_displacement,
                    self.max_displacement + 1,
                    device=input1.device,
                ),
                torch.arange(
                    -self.max_displacement,
                    self.max_displacement + 1,
                    device=input1.device,
                ),
            ),
            dim=-1,
        ).reshape(-1, 2)

        # Initialize the output tensor
        correlation = torch.zeros(
            b, offsets.shape[0], output_height, output_width, device=input1.device
        )

        # Compute the correlation using efficient tensor operations
        for idx, (dy, dx) in enumerate(offsets):
            shifted_patches2 = F.unfold(
                input2_padded[
                    :,
                    :,
                    dy + self.max_displacement : dy + self.max_displacement + h,
                    dx + self.max_displacement : dx + self.max_displacement + w,
                ],
                kernel_size=self.kernel_size,
                dilation=self.dilation,
                padding=self.padding,
                stride=self.stride,
            )
            shifted_patches2 = shifted_patches2.view(
                b,
                c * self.kernel_size[0] * self.kernel_size[1],
                output_height,
                output_width,
            )
            correlation[:, idx, :, :] = torch.mean(patches1 * shifted_patches2, dim=1)

        correlation = correlation.view(
            b, self.patch_size[0], self.patch_size[1], output_height, output_width
        )

        return correlation
```

I've not fully tested the numerically correctness. 

@jansel It compiles correctly with `@torch.compile` but with `@torch.compile(fullgraph=True)` it fail on:

```python
                input2_padded[
                    :,
                    :,
                    dy + self.max_displacement : dy + self.max_displacement + h,
                    dx + self.max_displacement : dx + self.max_displacement + w,
                ],
```

Cause
```python
torch._dynamo.exc.Unsupported: Dynamic slicing on data-dependent value is not supported
```

**Comment by bhack** [comment]
@ezyang Do we need to inject `torch._check / torch:_dynamic_check` to make this `fullgraph` compatible?

**Comment by ezyang** [comment]
No, looking at the code, we just need to actually implement this codepath

**Comment by ezyang** [comment]
file a separate issue for this ig

**Comment by bhack** [comment]
@ezyang Do you think that I could add this example on a already open ticket like:
https://github.com/pytorch/pytorch/issues/127154
https://github.com/pytorch/pytorch/issues/123592

Or do you prefer a new one?

**Comment by ezyang** [comment]
oh, the old issue is good
